systemd/systemd.go

```diff
@@ -939,57 +975,121 @@ func (s *systemd) IsActive(serviceName string) (bool, error) {
+		// change of 'notifyEnable', which is detected by 'notifyShowFirst')
+		notifyEnable := false
+		notifyShowFirst := false
+
+		// Once the async systemd stop completes, we do one last status poll
+		// to make sure we have the most up to date state.
+		stopComplete := false
+
+		for {
+			select {
+			case <-quit:
+				// We are now complete, but poll final state of units
+				stopComplete = true
+
+			case <-check.C:
+				// Enable state poll of units
+
+			case <-notify.C:
+				// Enable user notifications and trigger first message
+				notifyEnable = true
+				notifyShowFirst = true
+			}
+
+			// Check if any of the remaining running units have stopped?
 			stillRunningServices := []string{}
 			for _, service := range serviceNames {
 				bs, err := s.systemctl("show", "--property=ActiveState", service)
 				if err != nil {
-					return err
+					errorRet <- err
+					return
 				}
 				if !isStopDone(bs) {
 					stillRunningServices = append(stillRunningServices, service)
-					allStopped = false
 				}
 			}
-			if allStopped {
-				return nil
+
+			// Any remaining running units stopped?
+			somethingChanged := len(serviceNames) != len(stillRunningServices)
+
+			// We only poll units still running.
+			serviceNames = stillRunningServices
+
+			if len(serviceNames) == 0 || stopComplete {
+				// We return without explicitly writing to the error channel
+				// because the deferred close will cause the pending channel
+				// read to return a nil error.
+				return
 			}
-			if !firstCheck {
-				// do not notify about services waiting on the
-				// first pass
-				continue loop
+
+			// The first time the notification silence period expires we print
+			// an update, or on any subsequent change to the list of waiting units.
+			if (notifyEnable && somethingChanged) || notifyShowFirst {
+				s.reporter.Notify(fmt.Sprintf("Waiting for %s to stop.", strutil.Quoted(serviceNames)))
+				notifyShowFirst = false
 			}
-			serviceNames = stillRunningServices
-			firstCheck = false
-		case <-notify.C:
 		}
-		// after notify delay or after a failed first check
-		s.reporter.Notify(fmt.Sprintf("Waiting for %s to stop.", strutil.Quoted(serviceNames)))
+	}()
+
+	// This command blocks until the 'systemctl stop' completes
+	_, errStop := s.systemctl(append([]string{"stop"}, serviceNames...)...)
+
+	// Notify the progress loop to exit since systemctl completed the request
+	close(quit)
+
+	// Wait until the progress loop returns
+	errProgress := <-errorRet
+
+	// If we have an error from 'systemctl stop', return this as first priority
+	if errStop != nil {
+		return errStop
+	}
+
+	// If we have an error from 'systemctl show', return this as second priority
+	if errProgress != nil {
+		return errProgress
 	}
 
-	return &Timeout{action: "stop", services: serviceNames}
+	// No error, but not all units have stopped
+	if len(serviceNames) != 0 {
+		return &Timeout{action: "stop", services: serviceNames}
```

Comment:
Should this still be a timeout error now that the function does no longer take a timeout?

Reply:
addressed above.